### PR TITLE
Processing Filter Toggle Mode

### PR DESCRIPTION
**Pull request by SteveMacenski**
In response to discussion in https://github.com/cra-ros-pkg/robot_localization/issues/416

Gives a service interface to pause accepting and processing new measurements but continuing to provide pose estimate information at the requested rate.

### CODE REVIEW

**Comment by SteveMacenski** [comment]
Toggling CI

File: include/robot_localization/ros_filter.h [hunk 1]
```diff
@@ -610,6 +627,9 @@ template<class T> class RosFilter
     //! @brief Whether the filter is enabled or not. See disabledAtStartup_.
     bool enabled_;
 
+    //! $brief Whether the filter should process new measurements or not.
+    bool toggled_on_;
```
Comment (ayrton04) [inline include/robot_localization/ros_filter.h]: As much as I hate it, this should be `toggledOn_`.

**Review by ayrton04 — COMMENTED** [review]
This looks good. I do have one concern, though, which is that a large "toggle off" time will result in a huge time delta for the first prediction step that happens when the node is toggled back on. The process noise is [multiplied by the time delta](https://github.com/cra-ros-pkg/robot_localization/blob/kinetic-devel/src/ekf.cpp#L389) and then added to the state covariance during the prediction step, so your prediction step will have a massive covariance increase, which will only be corrected for any measured variables in your first "toggle on" cycle. You might want to verify this by looking at the first EKF output message post-toggle-on.

I _think_ one easy way to fix this is to call `setLastMeasurementTime` with the current time after clearing the measurement queue in your "toggle off" case.

EDIT: my comments obviously only apply if you aren't using dynamic PNC.

File: include/robot_localization/ros_filter.h (hunk 1 above)
Comment (SteveMacenski) [reply include/robot_localization/ros_filter.h]: done

**Comment by SteveMacenski** [comment]
This is a good point, I'll have to think about this for a little while. That's probably the right answer but something about it feels 'wrong' to me.

Edit: after reading the filter base, I agree, this is the right move. commit to follow.

**Review by ayrton04 — APPROVED** [review]

**Comment by ayrton04** [comment]
Thanks for the PR!

**Comment by Timple** [comment]
Would it make sense here to set the initial state of the toggle with a parameter?

**Comment by ayrton04** [comment]
As in, a ROS parameter? There's already a parameter for that behavior. It's `disabled_at_startup`.

https://github.com/cra-ros-pkg/robot_localization/blob/c8b8e3f2ba478a6301cf0a9e6f2c9c3c6896f372/src/ros_filter.cpp#L915

The filter starts in a disabled state, and requires a service call to make it start processing measurements and producing estimates.

It would make sense to unify them, really, but this at least provides the functionality you may be after.

**Comment by Timple** [comment]
That is exactly what I was looking for. I seemed to have missed that while looking at the diff.

Thanks!

**Comment by ayrton04** [comment]
It's been around for a while, but it occurs to me that I may not have put it in the documentation. Sorry about that!

**Pull request merged**